### PR TITLE
Encodes and fetch the last election certificate IDs in the batch header

### PR DESCRIPTION
**Pull request by howardwu**
## Motivation

This PR encodes and fetch the last election certificate IDs in the batch header, with backwards compatibility.

The sister PR for this is here: https://github.com/AleoHQ/snarkVM/pull/2204

### CODE REVIEW

File: node/bft/src/bft.rs [hunk 1]
```diff
@@ -554,6 +566,12 @@ impl<N: Network> BFT<N> {
                 dag_write.commit(certificate, self.storage().max_gc_rounds());
             }
         }
+        // Update the last election certificate IDs.
+        {
+            let mut last_election_certificate_ids = self.last_election_certificate_ids.write();
+            last_election_certificate_ids.clear();
+            last_election_certificate_ids.extend(election_certificate_ids);
```
Comment (ljedrz) [inline node/bft/src/bft.rs]:
```suggestion
            *last_election_certificate_ids = election_certificate_ids;
```
this should be possible

File: node/bft/src/bft.rs [hunk 2]
```diff
@@ -724,14 +754,25 @@ impl<N: Network> BFT<N> {
...
+            for (leader_certificate, election_certificate_ids) in past_leader_certificates {
                 // Commit the leader certificate.
                 dag.commit(leader_certificate, self.storage().max_gc_rounds());
+                // Update the last election certificate IDs.
+                //
+                // Note: Because we will be committing the latest leader certificate after this,
+                // technically we do not need to be updating the last election certificate IDs
+                // for intermediate leader certificates. However, this is a safety mechanic to ensure completeness.
+                last_election_certificate_ids.clear();
+                last_election_certificate_ids.extend(election_certificate_ids);
```
Comment (ljedrz) [inline node/bft/src/bft.rs]:
```suggestion
                *last_election_certificate_ids = election_certificate_ids;
```

**Review by ljedrz — APPROVED** [review]

**Comment by niklaslong** [comment]
The code looks good to me at first glance, still need to test this locally. 🕵️

**Review by vvp — APPROVED** [review]
Looks ok :+1:  (though the code could use some comments on why the last election cert ids are needed).

File: node/bft/src/primary.rs [hunk 3]
```diff
@@ -1387,15 +1400,61 @@ impl<N: Network> Primary<N> {
...
+
+    /// Fetches any missing election certificates for the specified batch header from the specified peer.
+    async fn fetch_missing_election_certificates(
+        &self,
+        peer_ip: SocketAddr,
+        batch_header: &BatchHeader<N>,
+    ) -> Result<HashSet<BatchCertificate<N>>> {
+        // Retrieve the round.
+        let round = batch_header.round();
+        // If the previous round is 0, or is <= the GC round, return early.
```
Comment (acoglio) [inline node/bft/src/primary.rs:1434]:
```suggestion
        // If the previous round (i.e. round - 1, if round > 0) is 0, or is <= the GC round, return early.
```

**Review by acoglio — APPROVED** [review]

**Pull request merged**